conda/resolve.py

```diff
@@ -357,167 +325,181 @@ def verify_specs(self, specs):
         for ms in spec2:
             filter = self.default_filter(feats)
-            if not self.valid(ms, filter):
-                bad_deps.extend(self.invalid_chains(ms, filter))
+            bad_deps.extend(self.invalid_chains(ms, filter))
         if bad_deps:
             raise NoPackagesFoundError(bad_deps)
-        return spec2, opts, feats
+        return spec2, feats
+
+    def find_conflicts(self, specs):
+        """Perform a deeper analysis on conflicting specifications, by attempting
+        to find the common dependencies that might be the cause of conflicts.
+
+        Args:
+            specs: An iterable of strings or MatchSpec objects to be tested.
+            It is assumed that the specs conflict.
+
+        Returns:
+            Nothing, because it always raises an UnsatisfiableError.
+
+        Strategy:
+            If we're here, we know that the specs conflict. This could be because:
+            - One spec conflicts with another; e.g.
+                  ['numpy 1.5*', 'numpy >=1.6']
+            - One spec conflicts with a dependency of another; e.g.
+                  ['numpy 1.5*', 'scipy 0.12.0b1']
+            - Each spec depends on *the same package* but in a different way; e.g.,
+                  ['A', 'B'] where A depends on numpy 1.5, and B on numpy 1.6.
+            Technically, all three of these cases can be boiled down to the last
+            one if we treat the spec itself as one of the "dependencies". There
+            might be more complex reasons for a conflict, but this code only
+            considers the ones above.
+
+            The purpose of this code, then, is to identify packages (like numpy
+            above) that all of the specs depend on *but in different ways*. We
+            then identify the dependency chains that lead to those packages.
+        """
+        sdeps = {}
+        # For each spec, assemble a dictionary of dependencies, with package
+        # name as key, and all of the matching packages as values.
+        for ms in specs:
+            rec = sdeps.setdefault(ms, {})
+            slist = [ms]
+            while slist:
+                ms2 = slist.pop()
+                deps = rec.setdefault(ms2.name, set())
+                for fkey in self.find_matches(ms2):
+                    if fkey not in deps:
+                        deps.add(fkey)
+                        slist.extend(ms3 for ms3 in self.ms_depends(fkey) if ms3.name != ms.name)
+        # Find the list of dependencies they have in common. And for each of
+        # *those*, find the individual packages that they all share. Those need
+        # to be removed as conflict candidates.
+        commkeys = set.intersection(*(set(s.keys()) for s in sdeps.values()))
+        commkeys = {k: set.intersection(*(v[k] for v in sdeps.values())) for k in commkeys}
+        # and find the dependency chains that lead to them.
+        bad_deps = []
+        for ms, sdep in iteritems(sdeps):
+            filter = {}
```

Comment:
`filter` is a python builtin.  In general I **strongly** favor expressiveness over brevity, but here `filt` or `filter_map` might be more appropriate.  Especially because on line 401 `filter` almost looked like you were passing a callable.